components/calendar/src/any_calendar.rs

```diff
@@ -0,0 +1,324 @@
+            (&Self::Iso(ref $cal_matched), &AnyDateInner::Iso(ref $date_matched)) => $e,
+            _ => panic!(
+                "Found AnyCalendar with mixed calendar type {} and date type {}!",
+                $cal.calendar_name(),
+                $date.calendar_name()
+            ),
+        }
+    };
+}
+
+impl Calendar for AnyCalendar {
+    type DateInner = AnyDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> AnyDateInner {
+        match *self {
+            Self::Gregorian(ref c) => AnyDateInner::Gregorian(c.date_from_iso(iso)),
+            Self::Buddhist(ref c) => AnyDateInner::Buddhist(c.date_from_iso(iso)),
+            Self::Japanese(ref c) => AnyDateInner::Japanese(c.date_from_iso(iso)),
+            Self::Indian(ref c) => AnyDateInner::Indian(c.date_from_iso(iso)),
+            Self::Coptic(ref c) => AnyDateInner::Coptic(c.date_from_iso(iso)),
+            Self::Iso(ref c) => AnyDateInner::Iso(c.date_from_iso(iso)),
+        }
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        match_cal_and_date!(match (self, date): (c, d) => c.date_to_iso(d))
+    }
+
+    fn months_in_year(&self, date: &Self::DateInner) -> u8 {
+        match_cal_and_date!(match (self, date): (c, d) => c.months_in_year(d))
+    }
+
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        match_cal_and_date!(match (self, date): (c, d) => c.days_in_year(d))
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        match_cal_and_date!(match (self, date): (c, d) => c.days_in_month(d))
+    }
+
+    fn offset_date(&self, date: &mut Self::DateInner, offset: DateDuration<Self>) {
+        match (self, date) {
+            (&Self::Gregorian(ref c), &mut AnyDateInner::Gregorian(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (&Self::Buddhist(ref c), &mut AnyDateInner::Buddhist(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (&Self::Japanese(ref c), &mut AnyDateInner::Japanese(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (&Self::Indian(ref c), &mut AnyDateInner::Indian(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (&Self::Coptic(ref c), &mut AnyDateInner::Coptic(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (&Self::Iso(ref c), &mut AnyDateInner::Iso(ref mut d)) => {
+                c.offset_date(d, offset.cast_unit())
+            }
+            (_, d) => panic!(
+                "Found AnyCalendar with mixed calendar type {} and date type {}!",
+                self.calendar_name(),
+                d.calendar_name()
+            ),
+        }
+    }
+
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        calendar2: &Self,
```

Comment:
Suggestion (optional): Make this a single `Date` argument instead of the pair `DateInner+Calendar`?

Reply:
I wanted to, but Date abstracts over different kinds of ownership. I could use `Date<&C>` and construct one on the fly, but at that point it's basically the same thing.

These APis aren't intended to be directly called.

Reply:
I understand. It might be nicer to do `Date<&C>` so that it is self-contained but I understand that `Date` is a user-level type and it's fine to do something different down here in the guts.

Reply:
Yeah that was my thinking basically.